retrofit-converters/gson/src/main/java/retrofit/GsonConverter.java

```diff
@@ -15,71 +15,55 @@
  */
 package retrofit;
 
-import com.google.gson.Gson;
+import com.google.gson.TypeAdapter;
 import com.squareup.okhttp.MediaType;
 import com.squareup.okhttp.RequestBody;
 import com.squareup.okhttp.ResponseBody;
 import java.io.IOException;
-import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.lang.reflect.Type;
+import java.io.OutputStreamWriter;
+import java.io.Reader;
+import java.io.Writer;
 import java.nio.charset.Charset;
+import okio.Buffer;
 
-/**
- * A {@link Converter} which uses GSON for serialization and deserialization of entities.
- */
-public class GsonConverter implements Converter {
-  private final Gson gson;
+final class GsonConverter implements Converter {
+  private final TypeAdapter typeAdapter;
   private final Charset charset;
   private final MediaType mediaType;
 
-  /**
-   * Create an instance using a default {@link Gson} instance for conversion. Encoding to JSON and
-   * decoding from JSON (when no charset is specified by a header) will use UTF-8.
-   */
-  public GsonConverter() {
-    this(new Gson());
-  }
-
-  /**
-   * Create an instance using the supplied {@link Gson} object for conversion. Encoding to JSON and
-   * decoding from JSON (when no charset is specified by a header) will use UTF-8.
-   */
-  public GsonConverter(Gson gson) {
-    this(gson, Charset.forName("UTF-8"));
-  }
-
-  /**
-   * Create an instance using the supplied {@link Gson} object for conversion. Encoding to JSON and
-   * decoding from JSON (when no charset is specified by a header) will use the specified charset.
-   */
-  public GsonConverter(Gson gson, Charset charset) {
-    if (gson == null) throw new NullPointerException("gson == null");
-    if (charset == null) throw new NullPointerException("charset == null");
-    this.gson = gson;
+  GsonConverter(TypeAdapter typeAdapter, Charset charset, MediaType mediaType) {
+    this.typeAdapter = typeAdapter;
     this.charset = charset;
-    this.mediaType = MediaType.parse("application/json; charset=" + charset.name());
+    this.mediaType = mediaType;
   }
 
-  @Override public Object fromBody(ResponseBody body, Type type) throws IOException {
+  @Override public Object fromBody(ResponseBody body) throws IOException {
     Charset charset = this.charset;
     if (body.contentType() != null) {
       charset = body.contentType().charset(charset);
     }
 
-    InputStream is = body.byteStream();
+    Reader in = new InputStreamReader(body.byteStream(), charset);
```

Comment:
Can we get closer with `body.charStream()` ? It already uses the HTTP response's charset